added nested array support for valdiateEquals and validateDifferent

mangopeaches opened the pull request:
Hey there,

Great validation library, I really appreciate the usability and simplicity of it all! I've been using this for a bit and I've been sending it nested arrays decoded from my json api directly and I found that the equals and different methods weren't working for nested array elements. I made a quick fix that utilizes your existing helper function to extract the value from the second argument, as before it was using the name of the second item directly and therefore the isset($this->_fields[$field2]) would always evaluate to false when passing a nested array item (i.e. 'object.itemTwo'). I added four new tests that mimic your existing tests for these methods, but with nested arrays and everything looks good. I tested with more depth to the arrays and that appears to be fine as well. Let me know if you have any questions or comments. Thanks again for a great library!

Best,
Tom

willemwollebrants commented:
Thanks, this looks very useful. I wonder what we should do if both values are null. If I read the code correctly, values that aren't set will always return false, but if they are both not set/null they could be considered equal too I guess?

mangopeaches commented:
Hey,

Thanks for pointing that out. I added a few additional tests to account for those cases and found that the results line up with the existing behavior so this added functionality will not deviate from how it currently is performing. It turns out that in the case where both are null or not present in the input array and the field is not required that the validateEquals() and validateDifferent() method are never called because these fields are skipped over in the validate() method in the condition on line 955. Therefore, in the case where both were null or not set it and not required it would always evaluate to true. It's really the presence of the required rule which dictates the outcome of the overall validation outcome pertaining to this null/unset condition, which will always be false when either is null or not present, as expected. I ran these same additional unit tests that I have just added (minus the nested versions) with the previous code before this nested change and the single variable tests that I have added for these null/unset conditions performed the same as after this change, so I see this as an innocuous change which adds the nested support and is in line with the existing behavior for those two cases you have mentioned. Let me know if you notice anything else and hopefully this helps clarify.

Thanks,
Tom

willemwollebrants commented:
Thank you so much, both for the new functionality and the additional tests!
